design/ceph-mon-pv.md

```diff
@@ -0,0 +1,139 @@
+claim template that is used to obtain PVs for the monitor data store.
+
+```go
+type MonSpec struct {
+	Count                int  `json:"count"`
+	PreferredCount       int  `json:"preferredCount"`
+	AllowMultiplePerNode bool `json:"allowMultiplePerNode"`
+	VolumeClaimTemplate  *v1.PersistentVolumeClaim
+}
+```
+
+Unlike the storage requirements of OSDs (xf: StorageClassDeviceSets) it is
+reasonable to use default requirements for monitors (e.g. 5-10 GB). This means
+that a minimal claim may be specified using only the requested storage class in
+which case default resource requests will be used.
+
+```yaml
+volumeClaimTemplates:
+- spec:
+    storageClassName: gp2-ebs
+...
+volumeClaimTemplates:
+- spec:
+    storageClassName: local-storage
+```
+
+It is also reasonable for monitors to use storage in the form of a directory on
+the host's root file system. This behavior effectively provides something like
+Rook-specific dynamic HostPath provisioning and is the current default behavior
+in Rook. In order to continue to support this level of operational simplicity,
+the volume claim template itself can be left unspecified entirely to request the
+existing default behavior. Note that the storage class `local-storage` is not
+necessarily equivalent---there are a variety of implementations with the most
+popular requiring static provisioning [0, 1].
+
+[0]: https://github.com/rancher/local-path-provisioner
+[1]: https://github.com/kubernetes-sigs/sig-storage-local-static-provisioner
+
+## Non-PV monitor storage
+
+Not all monitor storage requirements should be satisfied using PV storage. Most
+notably log and crash data. These data may continue to be stored in HostPath
+volume mounts. This will continue to be based on `dataDirHostPath`.
+
+## Host networking
+
+Currently Ceph monitor pods may be deployed using host-based networking, or
+placed behind a service using a portable network address. This poses a challenge
+when migrating a monitor (and it's PV) to a new node which has been deployed
+using host-based networking as the monitor's address will no longer valid.
+
+Support for migration of monitors with host-based networking is future work. To
+simplify the initial implementation, host-based networking will not be supported
+for non-local PVs.
+
+# StatefulSet pod management
+
+Monitor pods will be managed by a StatefulSet rather than Deployments which are
+currently use. This change offers (1) simplified naming for DNS and pods (e.g.
+`ceph-mon-a` rather than `ceph-mon-a-3ab423fb`), and (2) StatefulSets are able
+to handle much of the complexity of managing PVs including the mechanics of
+PV/pod association and scheduling.
+
+StatefulSets will be used to manage monitors pods in a 1-1 mapping between a pod
+and StatefulSet. This choice is made to retain maximum flexibility on naming,
+placement, and life cycle management of the pods by Rook.
+
+## Scheduling
+
+Rook currently makes explicit scheduling decisions for monitors. Standard node
```

Comment:
If we are using StatefulSets, Rook's current mon scheduling wouldn't apply, right? Because k8s then does the scheduling.

Reply:
I'm still a little not fully certain about all the scenarios regarding scheduling. I've experimented with some PoCs where I provisioned local PVs where I wanted the pod to run, rather than scheduling the pod itself. For non-local PVs I think we can apply a node selector to the statefulset and rely on the portability of non-local PVs. I'm still looking into some of the details around this. Do you have a specific case in mind?

Reply:
Not exactly. Perhaps I'm misunderstanding a bit; is one of your ideas to use one stateful set for each mon so that the mon's PV is handled by kubernetes?

Reply:
The design calls for one StatefulSet per monitor (i.e. replicas = 1). For non-local PVs, it is sufficient to set a node selector on the StatefulSet pod spec to achieve scheduling, so this is fully compatible with the existing scheduler code we have in rook for monitor placement. For local PVs we can use the same scheduling code, it just might be predicated on the availability of local PVs having been provisioned.